### PR TITLE
threefry #2418

### PR DESCRIPTION
**Pull request by gabelerner**
https://github.com/tinygrad/tinygrad/issues/2418

On metal (bigger shape breaks it because it's not efficient yet and runs out of buffers):
image

On CPU:
image

### CODE REVIEW

File: tinygrad/tensor.py [hunk 1]
```diff
@@ -153,12 +153,39 @@ def empty(*shape, **kwargs):
     return Tensor._loadop(LoadOps.EMPTY, prod((shape:=argfix(*shape))), **kwargs).reshape(shape)
 
   _seed: int = int(time.time())
+  _rng_counter: int = 0 # TODO: this has to be per device
   @staticmethod
   def manual_seed(seed=0): Tensor._seed = seed
 
+  @staticmethod
+  def _rand(): # threefry algo
+    key = [Tensor._seed, 0x0] # TODO: replace 0x0 with the CPU/GPU device this is running on
```
Comment (gabelerner) [inline tinygrad/tensor.py:162]: the `0x0` should be the device the tensor is running on. from my reading, it doesn't really matter which of the 4 values of key/x you change, but if all 4 are the same you get the same output

File: tinygrad/tensor.py [hunk 2]
```diff
@@ -153,12 +153,39 @@ def empty(*shape, **kwargs):
...
   _seed: int = int(time.time())
+  _rng_counter: int = 0 # TODO: this has to be per device
   @staticmethod
   def manual_seed(seed=0): Tensor._seed = seed
 
+  @staticmethod
+  def _rand(): # threefry algo
+    key = [Tensor._seed, 0x0] # TODO: replace 0x0 with the CPU/GPU device this is running on 
+    x = Tensor([Tensor._rng_counter, 0x0], dtype=dtypes.uint32)
+    Tensor._rng_counter += 1
```
Comment (gabelerner) [inline tinygrad/tensor.py:164]: increment just current devices rng counter

File: tinygrad/tensor.py [hunk 3]
```diff
@@ -153,12 +153,39 @@ def empty(*shape, **kwargs):
     return Tensor._loadop(LoadOps.EMPTY, prod((shape:=argfix(*shape))), **kwargs).reshape(shape)
 
   _seed: int = int(time.time())
+  _rng_counter: int = 0 # TODO: this has to be per device
```
[the remark this answers is not in the document]
Comment (blueridanus) [reply tinygrad/tensor.py:156]: maybe I'm misinterpreting, but moving tensors across devices is already possible fiy, e.g.: `tensor.to("gpu:2")`

[the remark this answers is not in the document]
Comment (gabelerner) [reply tinygrad/tensor.py:156]: sorry i think i miscommunicated, moving to the nth device. like if you have 4 GPUs and you want to pin this to the 2nd one

[the remark this answers is not in the document]
Comment (gabelerner) [reply tinygrad/tensor.py:156]: although based on your comment if `to` accepts `gpu:2` i can try to implement the map in this PR! thanks for letting me know about this.

**Pull request closed without merging**